File: javascript/customer-success-balancing.js
const CUSTOMERS_SUCCESS_WITH_SAME_AMOUT_OF_CUSTOMERS = 0;

/**
 * Returns the id of the CustomerSuccess with the most customers
 * @param {array} customerSuccess
 * @param {array} customers
 * @param {array} customerSuccessAway
 */

function customerSuccessBalancing(
  customerSuccess,
  customers,
  customerSuccessAway
) {
  const customersSuccessAvailables = getAvaliablesCustomersSuccess(
    customerSuccess,
    customerSuccessAway
  );
  const customersSuccessAvailablesSortedByScore = sortCustomersSuccessByScore(
    customersSuccessAvailables
  );

  const customersGroupedByScore = groupCustomersByScore(customers);

  const customersSuccessWithCustomers =
    addCustomersToQualifiedsCustomersSuccess(
      customersSuccessAvailablesSortedByScore,
      customersGroupedByScore
    );

  const customersSuccessSortedByHowManyCustomersCanServe =
    sortCustomersSuccessByWichCustomersCanServe(customersSuccessWithCustomers);

  if (
    customersSuccessSortedByHowManyCustomersCanServe[0].customers.length !==
    customersSuccessSortedByHowManyCustomersCanServe[1].customers.length
  ) {
    return customersSuccessSortedByHowManyCustomersCanServe[0].id;
  }

  return CUSTOMERS_SUCCESS_WITH_SAME_AMOUT_OF_CUSTOMERS;
}

function getAvaliablesCustomersSuccess(customerSuccess, customerSuccessAway) {
  return customerSuccess.filter((cs) => !customerSuccessAway.includes(cs.id));
}

function sortCustomersSuccessByScore(customersSuccess) {
  return customersSuccess.sort((cs1, cs2) => cs1.score - cs2.score);
}

function sortCustomersSuccessByWichCustomersCanServe(customersSuccess) {
  return customersSuccess.sort(
    (cs1, cs2) => cs2.customers.length - cs1.customers.length
  );
}

function groupCustomersByScore(customers) {
  return customers.reduce((customersGroupedByScore, customer) => {
    const groupOfCustomers = customersGroupedByScore[customer.score];

    const newGroupOfCustomers = groupOfCustomers ?? [customer];

    customersGroupedByScore[customer.score] = newGroupOfCustomers;

    return customersGroupedByScore;
  }, {});
}

function addCustomersToQualifiedsCustomersSuccess(
  customersSuccess,
  customersGroupedByScore
) {
  const customersGroupsScores = Object.keys(customersGroupedByScore);

  return customersSuccess.map((cs) => {
    const csScore = cs.score;

    const firstOverScore = customersGroupsScores.findIndex(
      (value) => Number(value) > csScore
    );

    const customersGroupsScoresKeys = customersGroupsScores.splice(
      0,
      firstOverScore < 0 ? customersGroupsScores.length : firstOverScore
    );

    cs.customers = customersGroupsScoresKeys
      .map((scoreKey) => customersGroupedByScore[scoreKey])
      .flat();
    return cs;
  });
}

test("Scenario 1", () => {
  const css = [
    { id: 1, score: 60 },
    { id: 2, score: 20 },
    { id: 3, score: 95 },
    { id: 4, score: 75 },
  ];
  const customers = [
    { id: 1, score: 90 },
    { id: 2, score: 20 },
    { id: 3, score: 70 },
    { id: 4, score: 40 },
    { id: 5, score: 60 },
    { id: 6, score: 10 },
  ];
  const csAway = [2, 4];

  expect(customerSuccessBalancing(css, customers, csAway)).toEqual(1);
});

function buildSizeEntities(size, score) {
  const result = [];
  for (let i = 0; i < size; i += 1) {
    result.push({ id: i + 1, score });
  }
  return result;
}

function mapEntities(arr) {
  return arr.map((item, index) => ({
    id: index + 1,
    score: item,
  }));
}

function arraySeq(count, startAt) {
  return Array.apply(0, Array(count)).map((it, index) => index + startAt);
}

test("Scenario 2", () => {
  const css = mapEntities([11, 21, 31, 3, 4, 5]);
  const customers = mapEntities([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]);
  const csAway = [];

  expect(customerSuccessBalancing(css, customers, csAway)).toEqual(0);
});

test("Scenario 3", () => {
  const testTimeoutInMs = 100;
  const testStartTime = new Date().getTime();

  const css = mapEntities(arraySeq(999, 1));
  const customers = buildSizeEntities(10000, 998);
  const csAway = [999];

  expect(customerSuccessBalancing(css, customers, csAway)).toEqual(998);

  if (new Date().getTime() - testStartTime > testTimeoutInMs) {
    throw new Error(`Test took longer than ${testTimeoutInMs}ms!`);
  }
});

test("Scenario 4", () => {
  const css = mapEntities([1, 2, 3, 4, 5, 6]);
  const customers = mapEntities([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]);
  const csAway = [];

  expect(customerSuccessBalancing(css, customers, csAway)).toEqual(0);
});

test("Scenario 5", () => {
  const css = mapEntities([100, 2, 3, 6, 4, 5]);
  const customers = mapEntities([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]);
  const csAway = [];

  expect(customerSuccessBalancing(css, customers, csAway)).toEqual(1);
});

test("Scenario 6", () => {
  const css = mapEntities([100, 99, 88, 3, 4, 5]);
  const customers = mapEntities([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]);
  const csAway = [1, 3, 2];

  expect(customerSuccessBalancing(css, customers, csAway)).toEqual(0);
});

test("Scenario 7", () => {
  const css = mapEntities([100, 99, 88, 3, 4, 5]);
  const customers = mapEntities([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]);
  const csAway = [4, 5, 6];

  expect(customerSuccessBalancing(css, customers, csAway)).toEqual(3);
});

File: ruby/customer_success_balancing.rb
require 'minitest/autorun'
require 'timeout'

CUSTOMERS_SUCCESS_WITH_SAME_AMOUT_OF_CUSTOMERS = 0

class CustomerSuccessBalancing
  def initialize(customer_success, customers, away_customer_success)
    @customer_success = customer_success
    @customers = customers
    @away_customer_success = away_customer_success
  end

  # Returns the ID of the customer success with most customers
  def execute
    customers_success_availables = get_avaliables_customers_success()

    customers_success_availables_sorted_by_score = sort_customers_success_by_score(customers_success_availables)

    customers_grouped_by_score = group_customers_by_score()

    customers_success_with_customers = add_customers_to_qualifieds_customers_success(customers_success_availables_sorted_by_score, customers_grouped_by_score)

    customers_success_sorted_by_home_many_customers_can_serve = sort_customers_success_by_wich_customers_can_serve(customers_success_with_customers)

    if customers_success_sorted_by_home_many_customers_can_serve[0][:customers].length != customers_success_sorted_by_home_many_customers_can_serve[1][:customers].length
      return customers_success_sorted_by_home_many_customers_can_serve[0][:id]
    end

    return CUSTOMERS_SUCCESS_WITH_SAME_AMOUT_OF_CUSTOMERS
  end


  def get_avaliables_customers_success
    return @customer_success.filter{|customer_success| !@away_customer_success.include?(customer_success[:id])}
  end

  def sort_customers_success_by_score(customers_success)
    return customers_success.sort_by{|customer_success| customer_success[:score]}
  end
  
  def group_customers_by_score()
    customers_grouped_by_score = {}
    @customers.each do |customer|
      group_of_customers = customers_grouped_by_score[customer[:score]]
      new_group_of_customers = group_of_customers.nil? ? [customer] : group_of_customers + [customer]
      customers_grouped_by_score[customer[:score]] = new_group_of_customers
    end
    customers_grouped_by_score.sort.to_h
  end

  def add_customers_to_qualifieds_customers_success(customers_success, customers_grouped_by_score)
    customers_groups_scores = customers_grouped_by_score.keys

    customers_success.map do |cs|
      cs_score = cs[:score]

      first_overscore = customers_groups_scores.find_index { |score| score.to_i > cs_score }

      customers_groups_scores_keys = customers_groups_scores.slice!(0, first_overscore || customers_groups_scores.length)

      cs[:customers] = customers_groups_scores_keys
                          .map { |score_key| customers_grouped_by_score[score_key] }
                          .flatten
    end

    return customers_success
  end

  def sort_customers_success_by_wich_customers_can_serve(customers_success)
    return customers_success.sort_by { |cs| cs[:customers].length }.reverse
  end
end

class CustomerSuccessBalancingTests < Minitest::Test
  def test_scenario_one
    balancer = CustomerSuccessBalancing.new(
      build_scores([60, 20, 95, 75]),
      build_scores([90, 20, 70, 40, 60, 10]),
      [2, 4]
    )
    assert_equal 1, balancer.execute
  end

  def test_scenario_two
    balancer = CustomerSuccessBalancing.new(
      build_scores([11, 21, 31, 3, 4, 5]),
      build_scores([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]),
      []
    )
    assert_equal 0, balancer.execute
  end

  def test_scenario_three
    balancer = CustomerSuccessBalancing.new(
      build_scores(Array(1..999)),
      build_scores(Array.new(10000, 998)),
      [999]
    )
    result = Timeout.timeout(1.0) { balancer.execute }
    assert_equal 998, result
  end

  def test_scenario_four
    balancer = CustomerSuccessBalancing.new(
      build_scores([1, 2, 3, 4, 5, 6]),
      build_scores([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]),
      []
    )
    assert_equal 0, balancer.execute
  end

  def test_scenario_five
    balancer = CustomerSuccessBalancing.new(
      build_scores([100, 2, 3, 6, 4, 5]),
      build_scores([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]),
      []
    )
    assert_equal 1, balancer.execute
  end

  def test_scenario_six
    balancer = CustomerSuccessBalancing.new(
      build_scores([100, 99, 88, 3, 4, 5]),
      build_scores([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]),
      [1, 3, 2]
    )
    assert_equal 0, balancer.execute
  end

  def test_scenario_seven
    balancer = CustomerSuccessBalancing.new(
      build_scores([100, 99, 88, 3, 4, 5]),
      build_scores([10, 10, 10, 20, 20, 30, 30, 30, 20, 60]),
      [4, 5, 6]
    )
    assert_equal 3, balancer.execute
  end

  private

  def build_scores(scores)
    scores.map.with_index do |score, index|
      { id: index + 1, score: score }
    end
  end
end
